src/docs/arch-concurrency.dox

```diff
@@ -0,0 +1,107 @@
+/*! @arch_page arch-concurrency WiredTiger concurrency management
+
+WiredTiger is a multi-threaded application and needs to manage concurrent access to shared memory
+locations. Many mechanisms are defined in WiredTiger to safely access shared memory locations. This
+page describes what those mechanisms are and the intended usage of them in the code base.
+
+@section volatile Volatile
+
+A controversial tool in the C toolkit. WiredTiger utilizes volatile for one reason, managing
+compiler optimization. Leveraging the definitions from "Is Parallel Programming Hard, And, If So,
+What Can You Do About It?", Paul E. McKenney, WiredTiger utilizes volatile to prevent the
+following type of compiler optimizations: Load fusing, store fusing, invented loads, invented
+stores and code reordering. WiredTiger defines the following macros WT_READ_ONCE and
+WT_WRITE_ONCE, these macros, if used correctly, prevent the described forms of compiler
+optimization. Additionally some variables in WiredTiger are defined with the volatile keyword if
+it is known that they will be frequently accessed in a concurrent context.
+
+@section locking Locking
+
+There are three types of locks in WiredTiger:
+ - Condition Variables
+ - Spin Locks
+ - Read Write (RW) Locks
+Each mechanism guarantees that once a thread relinquishes its lock and another thread obtains the
+same lock, all modifications made by the first thread will be visible to the thread now holding
+the lock.
+
+There are two implementations of condition variables in WiredTiger respectively for POSIX and MSVC
+under the same interface. The specific implementation is determined at compile time based on the
+platform.
+
+There are three different implementations for spin locks in WiredTiger under the same interface:
+ - An implementation using GCC's builtin atomic operations.
+ - An implementation using a POSIX based mutex.
+ - An implementation for MSVC.
+Again specific implementation is determined at compile time based on the platform and compile-time
+configurations. WiredTiger has its own read-write lock implementation. It is based on atomic
+operations and condition variables.
+
+@subsection lock-usage Lock usage
+A correctly used lock takes the lock for each access to the shared memory it guards. WiredTiger
+sometimes deviates from this model and reads variables outside the context of the relevant lock.
+Such instances are only permitted if reading inconsistent data is sufficient for the use case.
```

Comment:
I think this should be considered assumed reader knowledge. Multi-threaded programming is relatively commonplace.

Reply:
This speaks to a specific WiredTiger methodology, while I don't agree with our usage I was simply documenting it. As otherwise we have no documentation for this oddity.

Reply:
+1 
Having variables that are only sometimes protected by locks was enough of a surprise that I'd like to document it.

Reply:
IIRC we access without lock protection when it is thought that multi-threaded access is not possible due to the state of the system. I think explaining this reasoning is appropriate.

Reply:
Unfortunately that is not the case, the worst offender is the txn_global structure which has values read from it all over the place outside the txn_global rwlock. Two examples:

- https://github.com/wiredtiger/wiredtiger/blob/7facd2810f02f65be3a5530974aa3fc7b645daae/src/txn/txn_timestamp.c#L651
- https://github.com/wiredtiger/wiredtiger/blob/7facd2810f02f65be3a5530974aa3fc7b645daae/src/include/txn_inline.h#L660

Reply:
I agree with @luke-pearson and @ajmorton that it's worth specifically mentioning that variables that are only sometimes protected is a good idea. Otherwise readers of the code might incorrectly assume they've found a bug, or they might misunderstood how the code works in practice.